File: 백준/Silver/2156. 포도주 시식/포도주 시식.java
import java.io.*;
import java.util.*;

public class Main {
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		int n = Integer.parseInt(br.readLine());

		int[] wine = new int[n];

		for (int i = 0; i < n; i++) {
			wine[i] = Integer.parseInt(br.readLine());
		}

		int[] dp = new int[n + 1];

		if (n >= 1) dp[0] = wine[0];
		if (n >= 2) dp[1] = wine[0] + wine[1];
		if (n >= 3) dp[2] = Math.max(Math.max(wine[0] + wine[1], wine[0] + wine[2]), wine[1] + wine[2]);

		for (int i = 3; i < n; i++) {
			dp[i] = Math.max(Math.max(dp[i - 1], dp[i - 2] + wine[i]), dp[i - 3] + wine[i - 1] + wine[i]);
		}

		int max = 0;
		for (int i : dp) {
			max = Math.max(max, i);
		}
		
		System.out.println(max);
	}
}

File: 백준/Silver/2156. 포도주 시식/포도주 시식.py
import sys

input = sys.stdin.readline
n = int(input())
wine = []
for _ in range(n):
  w = int(input())
  wine.append(w) 

dp = [0] * n
if n == 1:
  print(wine[0])
elif n == 2:
  print(wine[0] + wine[1])
elif n == 3:
  print(max(wine[0] + wine[1], wine[1] + wine[2], wine[0] + wine[2]))
else:
  dp[0] = wine[0]
  dp[1] = wine[0] + wine[1]
  dp[2] = max(wine[2]+wine[1], wine[2]+wine[0], dp[1])

  for i in range(3, n):
    dp[i] = max(dp[i - 1], dp[i - 3] + wine[i - 1] + wine[i], dp[i - 2] + wine[i])
  print(max(dp))